posts/musings/2021-09-03-Components-for-Everyone.md

```diff
@@ -0,0 +1,94 @@
+
+**Elixir / Phoenix**
+
+The [`Phoenix.Component` API](https://hexdocs.pm/phoenix_live_view/Phoenix.Component.html#content) allows Elixir developers to create function components, similar to [React](https://reactjs.org/docs/components-and-props.html#function-and-class-components), including ability to pass in "assigns" (props) and "blocks" (children).
+
+[Surface UI](https://surface-ui.org/documentation) builds on top of the Phoenix component API to provide "a fast and productive solution to build modern web applications". It uses an extended templating syntax, helpful macros, and built-in components to become a more full-featured view framework for Phoenix applications.
+
+**JS / Node**
+
+Lastly, we get to the curious case of server-rendered component systems for NodeJS. The closest library in comparison to the previous examples is [Marko](https://markojs.com/), which allows developers to "start with simple HTML templates and add powerful components as needed". Marko templating can be integrated into [popular server libraries](https://markojs.com/docs/server-integrations-overview/) as the core view engine, rather than [forcing a typical client-side framework into that workflow](https://reactjs.org/docs/react-dom-server.html). There was an official view engine for [React in Express](https://github.com/reactjs/express-react-views), but it no longer appears actively supported; [React Server Components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html) are also still theoretical at the time of writing.
+
+There is active development in this space with the advent of tools like [Astro](https://astro.build/) and [Slinkity](https://slinkity.dev/) that aim to take a framework-agnostic approach to introduce components into modern, HTML-first static sites. Astro does a great job [comparing itself to existing tools](https://docs.astro.build/comparing-astro-vs-other-tools) and Slinkity is building on the shoulders of the awesome [Eleventy](https://www.11ty.dev/) ecosystem (slight bias given this site is currently built with Eleventy).
+
+Previously, it seemed like NodeJS devs didn't invest in a server-side component framework because existing client-side libraries could be worked into the SSR pipeline; why reinvent the wheel? There is less of a context switch in terms of syntax when compared to the workflows of other web languages, so the friction lies in the [uncanny vallye](https://crystallize.com/blog/frontend-performance-react-ssr-and-the-uncanny-valley) of managing a client-side component system in a traditional server application.
```

Comment:
```suggestion
Previously, it seemed like NodeJS devs didn't invest in a server-side component framework because existing client-side libraries could be worked into the SSR pipeline; why reinvent the wheel? There is less of a context switch in terms of syntax when compared to the workflows of other web languages, so the friction lies in the [uncanny valley](https://crystallize.com/blog/frontend-performance-react-ssr-and-the-uncanny-valley) of managing a client-side component system in a traditional server application.
```